idem_azurerm/states/azurerm/web/function_app.py

```diff
@@ -0,0 +1,571 @@
+                - resource_group: my_group
+                - functions_file_path: "/path/to/functions.zip"
+                - os_type: "Linux"
+                - runtime_stack: "python"
+                - storage_account: my_account
+                - app_service_plan: my_plan
+                - enable_app_insights: True
+                - tags:
+                    "Owner": "EITR Technologies"
+                - connection_auth: {{ profile }}
+
+    """
+    ret = {"name": name, "result": False, "comment": "", "changes": {}}
+    action = "create"
+    app_settings = [
+        {"name": "FUNCTIONS_WORKER_RUNTIME", "value": runtime_stack.lower()},
+        {
+            "name": "FUNCTIONS_EXTENSION_VERSION",
+            "value": ("~" + str(functions_version)),
+        },
+        {"name": "FUNCTION_APP_EDIT_MODE", "value": "readonly"},
+        {"name": "SCM_DO_BUILD_DURING_DEPLOYMENT", "value": "false"},
+    ]
+
+    if not isinstance(connection_auth, dict):
+        if ctx["acct"]:
+            connection_auth = ctx["acct"]
+        else:
+            ret[
+                "comment"
+            ] = "Connection information must be specified via acct or connection_auth dictionary!"
+            return ret
+
+    # Ensures location is specified
+    if "location" not in kwargs:
+        rg_props = await hub.exec.azurerm.resource.group.get(
+            ctx, resource_group, **kwargs
+        )
+
+        if "error" in rg_props:
+            log.error("Unable to determine location from resource group specified.")
+            return {
+                "error": "Unable to determine location from resource group specified."
+            }
+        kwargs["location"] = rg_props["location"]
+
+    # Handle storage account validation
+    if not storage_rg:
+        storage_rg = resource_group
+
+    storage_acct = await hub.exec.azurerm.storage.account.get_properties(
+        ctx, name=storage_account, resource_group=storage_rg
+    )
+
+    if "error" in storage_acct:
+        log.error(
+            "The storage account does not exist within the specified resource group."
+        )
+        ret[
+            "comment"
+        ] = "The storage account does not exist within the specified resource group."
+        return ret
+
+    # Ensure that the file path contains a zip file
+    filename = os.path.basename(functions_file_path)
+    if not ".zip" in filename.lower():
+        log.error(
+            "The specified file in functions_file_path is not a compressed (zip) file."
+        )
+        ret[
+            "comment"
+        ] = "The specified file in functions_file_path is not a compressed (zip) file."
+        return ret
+
+    # Set reserved for the ASP and Function App based upon OS type
+    if os_type.lower() == "windows":
+        reserved = False
+    else:  # linux
+        reserved = True
+
+    # Handle App Service Plan creation
+    if not app_service_plan:
+        app_service_plan = f"ASP-{name}"
```

Comment:
do we want to name it `plan-*`?
https://docs.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/naming-and-tagging#compute-and-web

Reply:
Yes I can swap it to `plan-*`. I had named it `ASP-*` because that is what Azure does when automatically creating it for the Function App through the portal. However, following Azure naming convention would probably be best.